docs/Developers/NextGen_Core.rst

```diff
@@ -0,0 +1,198 @@
+subclass of the ``PackageRequirement`` class for that package manager. For
+example in case of a pip requirement do the following:
+
+::
+
+    >>> from dependency_management.requirements.PipRequirement import (
+    ...     PipRequirement)
+    >>> class FooBear(Bear):
+    ...     REQUIREMENTS = {PipRequirement('coala_decorators', '0.2.1')}
+
+
+Override bears and define task scheduling
+-----------------------------------------
+
+A nextgen bear has to have the following functions to perform analysis
+- ``generate_tasks``: This method is a part of the parent ``Bear.py`` class
+and returns tuples containing the positional arguments as a tuple and the
+keyword arguments in the form of a dict. These are actually the task objects
+that are scheduled and executed by the core. Absence of this method raises
+``NotImplementedError``.
+- ``analyze``: This method contains the code that performs the actual code
+analysis routine that that bear is used for. Absence of this method raises
+``NotImplementedError``.
+
+How to use `Core.run`
+--------------------
+
+The `run(bears, result_callback, cache=None, executor=None)` method in `Core.PY`
+is used to instantiate a coala session with the parameters it receives and run
+it.
+- ``bears``: contains all the instances of bears to run.
+- ``result_callback``: a function that is called only when the results from
+the bears are available.
+- ``cache``: a cache used by bears to speed up runs.
+- ``executor``: a custom executor used to run the bears. By default
+``ProcessPoolExecutor`` is used.
+
+The new core comes with the following new additions:
+- **Dependency Resolution:** The ``initialize_dependencies`` method in
+``Core.py`` recieves the bears that are to be run and processes bear
+dependencies using a consumer based system so that each dependency bear has only
+one instance per section and file-dict. It returns a set of dependency bears
+along with those bears that don't have any dependencies or whose dependencies
+have been resolved(these are the ones that are scheduled to be run). The bears
+are resolved by traversing each edge of the dependency graph and instantiating
+dependency bears accordingly one per section. Before the bears are run we
+initialize the dependency tracking in the ``__init__`` method of the class
+``Session`` which is responsible for running coala sessions. The
+``dependency_tracker`` object is used while scheduling and cleaning up bears and
+also while finishing bear tasks.
+
+- **Task Scheduling:** The bears that have no dependencies or whose
+dependencies have been resolved, only their tasks will be scheduled for
+execution. Before exexuting any tasks the coala looks up the task in the cache.
+In case of a hit the existing results that is stored in the cache for the
+corresponding task arguments is called using ``execute_task_with_cache``
+method. In case of a miss or if coala is ran without a cache the bear task is
```

Comment:
decide if we call it a "task" or "bear task"

Reply:
that was confusing, switching it to task :+1: